### PR TITLE
Add work apply controller

### PR DESCRIPTION
**Pull request by qiujian16**
/hold

still need to add ut at least

### CODE REVIEW

**Comment by RainbowMango** [comment]
Good job!!! @qiujian16 
I'll look at it ASAP.

File: pkg/controllers/apply_controller.go [hunk 1]
```diff
@@ -0,0 +1,356 @@
...
+	identifier workv1alpha1.ResourceIdentifier
+	generation int64
+	updated    bool
+	err        error
+}
+
+// Reconcile implement the control loop logic for Work object.
+func (r *ApplyWorkReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	work := &workv1alpha1.Work{}
+	err := r.client.Get(ctx, types.NamespacedName{Name: req.Name, Namespace: req.Namespace}, work)
```
Comment (RainbowMango) [inline pkg/controllers/apply_controller.go:58]:
```suggestion
	err := r.client.Get(ctx, req.NamespacedName, work)
```
We don't need to rebuild a NamespaceName here.

File: pkg/controllers/apply_controller.go [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+	found := false
+	for i := range work.Finalizers {
+		if work.Finalizers[i] == workFinalizer {
+			found = true
+			break
+		}
+	}
+	if !found {
+		return ctrl.Result{}, nil
+	}
```
Comment (RainbowMango) [inline pkg/controllers/apply_controller.go:77]:
```suggestion
	if !controllerutil.ContainsFinalizer(work, workFinalizer) {
		return ctrl.Result{}, nil
	}
```
Maybe we can use [ContainsFinalizer](https://github.com/kubernetes-sigs/controller-runtime/blob/197751df6040ec99414574e89f3fa73914ce335d/pkg/controller/controllerutil/controllerutil.go#L363) to simply this logic.

File: pkg/controllers/apply_controller_test.go [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/runtime"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilrand "k8s.io/apimachinery/pkg/util/rand"
+	workv1alpha1 "sigs.k8s.io/work-api/pkg/apis/v1alpha1"
```
Comment (RainbowMango) [inline pkg/controllers/apply_controller_test.go:33]: Do you want to group imports?

File: pkg/controllers/finalize_controller.go [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+	client             client.Client
+	spokeDynamicClient dynamic.Interface
+	restMapper         meta.RESTMapper
+	log                logr.Logger
+}
+
+// Reconcile implement the control loop logic for finalizing Work object.
+func (r *FinalizeWorkReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	work := &workv1alpha1.Work{}
+	err := r.client.Get(ctx, types.NamespacedName{Name: req.Name, Namespace: req.Namespace}, work)
```
Comment (RainbowMango) [inline pkg/controllers/finalize_controller.go:43]:
```suggestion
	err := r.client.Get(ctx, req.NamespacedName, work)
```

File: pkg/controllers/finalize_controller.go [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+	if !work.DeletionTimestamp.IsZero() {
+		// TODO add clean resource logic
+		newFinalizers := []string{}
+		for i := range work.Finalizers {
+			if work.Finalizers[i] == workFinalizer {
+				continue
+			}
+			newFinalizers = append(newFinalizers, work.Finalizers[i])
+		}
+		work.Finalizers = newFinalizers
```
Comment (RainbowMango) [inline pkg/controllers/finalize_controller.go:61]:
```suggestion
		if controllerutil.ContainsFinalizer(work, workFinalizer) {
			controllerutil.RemoveFinalizer(work, workFinalizer)
		}
```
Does this work too?

**Review by RainbowMango — COMMENTED** [review]
Just some nits, otherwise LGTM.
Great job! Thanks again! @qiujian16

File: pkg/controllers/apply_controller_test.go (hunk 3 above)
Comment (qiujian16) [reply pkg/controllers/apply_controller_test.go]: updated

File: pkg/controllers/apply_controller.go (hunk 1 above)
Comment (qiujian16) [reply pkg/controllers/apply_controller.go]: updated

File: pkg/controllers/apply_controller.go (hunk 2 above)
Comment (qiujian16) [reply pkg/controllers/apply_controller.go]: yeah, this is neat.

**Comment by RainbowMango** [comment]
/lgtm
Left approval to @pmorie .

**Comment by RainbowMango** [comment]
/approve 
Let's move forward faster and iterate it later if there are any comments/issues.

**Pull request merged**